File: go/14-longest-collatz-sequence.go
package main

import "fmt"

func longestCollatzSequence(limit int) int {
	maxLen := 0
	maxNumber := 0

	limit--
	for limit > 0 {
		currNumber := limit
		currLen := 1
		for currNumber != 1 {
			if currNumber%2 == 0 {
				currNumber /= 2
			} else {
				currNumber = 3*currNumber + 1
			}
			currLen++
		}

		if currLen > maxLen {
			maxLen = currLen
			maxNumber = limit
		}

		limit--
	}

	return maxNumber
}

func main() {
	fmt.Println(longestCollatzSequence(1000000))
}

File: javascript/14-longest-collatz-sequence.js
// TODO: cache intermediate results for better performance
function longestCollatzSequence(limit) {
  let maxLen = 0;
  let maxNumber;

  limit--;
  while (limit > 0) {
    let currNumber = limit;
    let currLen = 1;
    while (currNumber !== 1) {
      if (currNumber % 2 === 0) {
        currNumber /= 2;
      } else {
        currNumber = 3 * currNumber + 1;
      }
      currLen++;
    }

    if (currLen > maxLen) {
      maxLen = currLen;
      maxNumber = limit;
    }

    limit--;
  }

  return maxNumber;
}

console.log(longestCollatzSequence(1000000));
